src/EditorFeatures/Core/InheritanceMargin/InheritanceMarginServiceHelpers.cs

```diff
@@ -124,68 +126,92 @@ internal static class InheritanceMarginServiceHelper
+                        lineNumber,
+                        implementingMembers: implementingSymbols,
+                        cancellationToken).ConfigureAwait(false);
+                    builder.AddIfNotNull(item);
+                }
+            }
+            else
+            {
+                // Go down the inheritance chain to find all the overriding targets.
+                var allOverridingSymbols = await SymbolFinder.FindOverridesArrayAsync(memberSymbol, solution, cancellationToken: cancellationToken).ConfigureAwait(false);
 
-            // Go down the inheritance chain to find all the implementing targets.
-            var allImplementedSymbols = await GetImplementedSymbolsForTypeMemberAsync(solution, memberSymbol, cancellationToken).ConfigureAwait(false);
+                // Go up the inheritance chain to find all overridden targets
+                var overriddenSymbols = GetOverriddenSymbols(memberSymbol);
 
-            // For all overriden & implemented symbols, make sure it is in source.
-            // For example, if the user is viewing System.Threading.SynchronizationContext from metadata,
-            // then don't show the derived overriden & implemented method in the default implementation for System.Threading.SynchronizationContext in metadata
-            var overriddenSymbols = allOverriddenSymbols.WhereAsArray(symbol => symbol.Locations.Any(l => l.IsInSource));
-            var implementedSymbols = allImplementedSymbols.WhereAsArray(symbol => symbol.Locations.Any(l => l.IsInSource));
+                // Go up the inheritance chain to find all the implemented targets.
+                var implementedSymbols = GetImplementedSymbolsForTypeMember(memberSymbol, overriddenSymbols);
 
-            if (overriddenSymbols.Any() || overridingSymbols.Any() || implementingSymbols.Any() || implementedSymbols.Any())
-            {
-                var item = await CreateInheritanceMemberInfoForMemberAsync(
-                    solution,
-                    memberSymbol,
-                    lineNumber,
-                    implementingMembers: implementingSymbols,
-                    implementedMembers: implementedSymbols,
-                    overridenMembers: overriddenSymbols,
-                    overridingMembers: overridingSymbols,
-                    cancellationToken).ConfigureAwait(false);
+                // For all overriding symbols, make sure it is in source.
+                // For example, if the user is viewing System.Threading.SynchronizationContext from metadata,
+                // then don't show the derived overriden & implemented method in the default implementation for System.Threading.SynchronizationContext in metadata
+                var overridingSymbols = allOverridingSymbols.WhereAsArray(symbol => symbol.Locations.Any(l => l.IsInSource));
 
-                builder.AddIfNotNull(item);
+                if (overridingSymbols.Any() || overriddenSymbols.Any() || implementedSymbols.Any())
```

Comment:
if CreateInheritanceMemberItemForClassOrStructMemberAsync already can return a nul value, then consider moving this code into it.